libsolidity/analysis/TypeChecker.cpp

```diff
@@ -596,19 +596,32 @@ bool TypeChecker::visit(VariableDeclaration const& _variable)
 	if (isStructMemberDeclaration)
 		return false;
 
-	if (auto referenceType = dynamic_cast<ReferenceType const*>(varType))
+	auto compositeType = dynamic_cast<CompositeType const*>(varType);
+	if (compositeType == nullptr)
+		return false;
+
+	bool callDataCheckRequired = true;
+	if (_variable.isLibraryFunctionParameter())
+		callDataCheckRequired = false;
```

Comment:
I think that all non-storage parameters in libraries should still have this check. They're all passed in `calldata`, like in normal external calls, so they should fit in it.

Storage parameters are an exception because we don't serialize the data from storage. We can safely pass a reference instead because library has access to the storage of the calling contract. Technically, we should validate the reference itself but it's so small that it will 100% pass the size check anyway.

Reply:
I added the check for storage location.